.github/workflows/docker-build-push.yml

```diff
@@ -21,20 +31,63 @@ jobs:
             type=ref,event=tag
 
       - name: Set up Docker Buildx
-        uses: docker/setup-buildx-action@v1.7.0
+        uses: docker/setup-buildx-action@v2.0.0
 
       - name: Login to DockerHub
         if: github.event_name != 'pull_request'
-        uses: docker/login-action@v1.14.1
+        uses: docker/login-action@v2.0.0
         with:
           username: ${{ secrets.DOCKERHUB_USERNAME }}
           password: ${{ secrets.DOCKERHUB_TOKEN }}
 
       - name: Build image and push to Docker Hub
-        uses: docker/build-push-action@v2.10.0
+        uses: docker/build-push-action@v3.0.0
         with:
           context: .
           file: docker/Dockerfile
           push: ${{ github.event_name != 'pull_request' }}
           tags: ${{ steps.docker_metadata.outputs.tags }}
           labels: ${{ steps.docker_metadata.outputs.labels }}
+
+  deploy_to_gce:
+    name: Deploy the image to a GCE VM
+    runs-on: ubuntu-latest
+    needs: push_to_registry
+
+    # Add "id-token" with the intended permissions.
+    permissions:
+      contents: "read"
+      id-token: "write"
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+
+      # Authentication via credentials json
+      - id: "auth"
+        uses: "google-github-actions/auth@v0"
+        with:
+          credentials_json: "${{ secrets.GCE_SA_KEY }}"
+
+      # Setup gcloud CLI
+      - name: Set up Cloud SDK
+        uses: google-github-actions/setup-gcloud@v0
+
+      # Make sure the VM is up
+      - name: Start the deploy-pudl-vm
+        run: gcloud compute instances start "$GCE_INSTANCE" --zone="$GCE_INSTANCE_ZONE"
+
+      # Deploy PUDL image to GCE
+      - name: Deploy
+        run: |-
+          gcloud compute instances update-container "$GCE_INSTANCE" \
+            --zone "$GCE_INSTANCE_ZONE" \
+            --container-image "docker.io/${{needs.push_to_registry.outputs.docker_tags}}" \
```

Comment:
What happens if there's more than one tag returned here? E.g. what if you're on a commit that's been tagged `v0.12.1` and also at the tip of the `dev` branch?

Reply:
I'm not sure. It seems like the Docker Metadata is only tagging the image with the branch name. Should we add commit sha too?

Reply:
But that's only because the branch isn't tagged. When we do a release we'll `git tag -s v0.9.0` and that will create another "reference" that is associated with a particular commit permanently, rather than following the `HEAD` of the branch.

I'd say go ahead and tag the branch just to see what happens. Maybe it will break things.

I don't understand the use case for tagging with the hash, though I've seen people do it.  You can always refer to the hash directly.

Reply:
Ah I see.  So if we run these builds on dev and main the dev image will always be `catalystcoop/pudl-etl:dev` and the main images will be tagged with our releases: `catalystcoop/pudl-etl:main-0.7.0`, `catalystcoop/pudl-etl:main-0.7.1`...

Reply:
I don't think that's how it works. You can tag the same image (the same hash) with many different tags, and they behave kind of like symlinks. E.g. I tag our `pudl-jupyter` images with the date, "latest" (if it's the latest image) and the hash (because I copied someone else's behavior.  So if you look over here, you'll see that all three of those point to the same actual image right now:

https://hub.docker.com/r/catalystcoop/pudl-jupyter

But they are distinct tags.

So we could have `catalystcoop/pudl-etl:dev` and also `catalystcoop/pudl-etl:v0.7.0` and have them both point the the same image, because it was both tagged `v0.7.0` and also happened, at that moment, to be derived from the tip of the `dev` branch. As soon as another commit is made and pushed and the image is rebuilt, the two tags would be on separate images, since `v0.7.0` only applies to a single unchanging commit, while `dev` is defined by being the HEAD of the `dev` branch, wherever that is from day to day.

When we're tagging the images, we can say "Tag this one image with these 63 different tags" and that's fine, but when we're pulling an image based on a tag, we need to specify only a single tag.

Reply:
Oooh, wait I see why it's useful to tag with the hash. Because it lets you link a particular image back to a specific individual commit in git.  Yeah, we should definitely do that.  I was confused because both the images and the commits are identified by hashes.

Reply:
Ahhh ok. I didn't realize you could create multiple tags for one image. Tagging the image with the commit hash sounds helpful.